dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/MessageExtensions.cs

```diff
@@ -0,0 +1,735 @@
+        /// <param name="handler">Function to call when the command is received.</param>
+        /// <returns>The application for chaining purposes.</returns>
+        public Application<TState, TTurnStateManager> OnBotMessagePreviewSend(Regex commandIdPattern, BotMessagePreviewSendHandler<TState> handler)
+        {
+            Verify.ParamNotNull(commandIdPattern);
+            Verify.ParamNotNull(handler);
+            RouteSelector routeSelector = CreateTaskSelector((string input) => commandIdPattern.IsMatch(input), SUBMIT_ACTION_INVOKE_NAME, "send");
+            return OnBotMessagePreviewSend(routeSelector, handler);
+        }
+
+        /// <summary>
+        /// Registers a handler to process the 'send' action of a message that's being previewed by the
+        /// user prior to sending.
+        /// </summary>
+        /// <param name="routeSelector">ID of the command to register the handler for.</param>
+        /// <param name="handler">Function to call when the command is received.</param>
+        /// <returns>The application for chaining purposes.</returns>
+        public Application<TState, TTurnStateManager> OnBotMessagePreviewSend(RouteSelector routeSelector, BotMessagePreviewSendHandler<TState> handler)
+        {
+            Verify.ParamNotNull(routeSelector);
+            Verify.ParamNotNull(handler);
+            RouteHandler<TState> routeHandler = async (ITurnContext turnContext, TState turnState, CancellationToken cancellationToken) =>
+            {
+                MessagingExtensionAction? messagingExtensionAction;
+                if (!string.Equals(turnContext.Activity.Type, ActivityTypes.Invoke, StringComparison.OrdinalIgnoreCase)
+                    || !string.Equals(turnContext.Activity.Name, SUBMIT_ACTION_INVOKE_NAME)
+                    || (messagingExtensionAction = Application<TState, TTurnStateManager>.GetInvokeValue<MessagingExtensionAction>(turnContext.Activity)) == null
+                    || !string.Equals(messagingExtensionAction.BotMessagePreviewAction, "send"))
+                {
+                    throw new TeamsAIException($"Unexpected MessageExtensions.OnBotMessagePreviewSend() triggered for activity type: {turnContext.Activity.Type}");
+                }
+
+                await handler(turnContext, turnState, messagingExtensionAction.BotActivityPreview.FirstOrDefault(), cancellationToken);
+
+                // Check to see if an invoke response has already been added
+                if (turnContext.TurnState.Get<object>(BotAdapter.InvokeResponseKey) == null)
+                {
+                    MessagingExtensionActionResponse response = new();
+                    Activity activity = Application<TState, TTurnStateManager>.CreateInvokeResponseActivity(response);
+                    await turnContext.SendActivityAsync(activity, cancellationToken);
+                }
+            };
+            _app.AddRoute(routeSelector, routeHandler);
```

Comment:
Is this an invoke handler, since the activity type is invoke?

Reply:
Yes, fixed.